### DISCUSSION TITLE
Hash Routing with Templates - Fail to make it work with some variation from the original

### DISCUSSION THREAD

**Discussion by RicardoFernandez-UY** [anvilistas/anvil-extras#454] [Q&A] [answered]
Hi,
I’m trying to create an application that:

Starts in a landing page (that currently just has a button to initiate)
Pressing that button takes you to a Public Area with some menu options and a login button.
**This works fine.**

Once you log into the application, you should be taken a different page (Template) that will have the menu options according to the area defined for your user, and the Forms corresponding to that options. The area in wich you work is an attribute of the user (unidad) and is taken from the users table. (Marketing is the only one implemented so far)
**This is the part that is not working fine**

3 Once you logout you should return to the landing page
**This apparently works fine**

For achieving this I:

Defined several Templates for Hash Routing to different groups of Forms mimicking the Hash routing with templates example. Mainly because the behavior I intend to have is very similar although it has some differences.

Each Template corresponds to a different area, and it has several associated Forms, and a different menu.

Once you login you should be directed to the corresponding Template that will call the Home Form for that Template (and not the Home of any other Template ), then generate and display the right menu for the user, that in turn will allow navigation to the rest of the Forms for that area.
I’ve done it up to here, but it is not working fine, so the problems are probably here

The menu is generated dinamically in each Template using Navigation module.
This works fine.

**Basically the problems starts when I log in and try to navigate to the Marketing area** (the only one I have defined so far, and the one the user belongs to). It then starts behaving oddly and unstable.

I’ve looked at the logs for the routing module when executing, and they show a behavior I can’t trace/understand. I’ve read the hash routing documentation but I can’t make a match for those error messages.

I think that the mechanism works well, and it’s me that have been unable to put the pieces in the right place and order.
Might be something related to the way I’m doing the navigation, or maybe some not so well defined decorators, but I’ve done many tries and can’t get it right.
The first thing you will probably notice is that the menu changes to the correct one, but the Home page does not corresponde to Marketing, and then when you click in any of the other links it crashes and returns to the Public Area??

**Here is the application. In case you want to test the behavior, the user for loggin in is [usuario@gmail.com] and the pass is uSuario-1**. This user belongs to the Marketing area, and so after login in should redirect to the Marketing menu, and be able to select the items from this menu and opens those Forms.

https://anvil.works/build#clone:MEZWXQDDZIV5RRLD=JUZZCTBVGTXKM6NEKAMHPSRO|C6ZZPAPN4YYF5NVJ=
I’ll really appreciate any help in order to understand where I’m messing things up. At this point I’ve completely lost the North.
Thank you very much!!

_**Below is the routing log that shows how once you are in the Marketing menu and click on a link (RedesSociales) it navigates back to the Public Template (I can't understand why), and throws an error because the /redesSociales path is not registered there, but in the Marketing Template. At this point I'm not able to map the code and the clicks with what the logs show**._

```
#routing: setting url_hash to: '#Marketing', adding to top of history stack
#routing: navigation triggered: url_hash='Marketing', url_pattern='Marketing', url_dict={}
#routing: entering navigation level: 0
  #routing: checking templates and redirects
  #routing: changing template: 'RouterAreaPublica' -> 'RouterMarketing'
  #routing: loaded template: 'RouterMarketing', re-navigating
  #routing: navigation triggered: url_hash='Marketing', url_pattern='Marketing', url_dict={}
  #routing: entering navigation level: 1
    #routing: checking templates and redirects
    #routing: unchanged template: 'RouterMarketing'
    #routing: RouterMarketing.on_navigation() called
    #routing: loading route: 'Home' from cache
  #routing: exiting navigation level: 1
#routing: exiting navigation level: 0
#routing: navigation complete
          
#routing: navigation triggered: url_hash='/redesSociales', url_pattern='/redesSociales', url_dict={}
#routing: entering navigation level: 0
  #routing: checking templates and redirects
  #routing: changing template: 'RouterMarketing' -> 'RouterAreaPublica'
  #routing: loaded template: 'RouterAreaPublica', re-navigating
  #routing: navigation triggered: url_hash='/redesSociales', url_pattern='/redesSociales', url_dict={}
  #routing: entering navigation level: 1
    #routing: checking templates and redirects
    #routing: unchanged template: 'RouterAreaPublica'
    #routing: RouterAreaPublica.on_navigation() called
    #routing: no route form with: url_pattern='/redesSociales' url_keys=[]template='RouterAreaPublica'
              If this is unexpected perhaps you haven't imported the form correctly
  #routing: exiting navigation level: 1
#routing: exiting navigation level: 0
#routing: navigation complete
          
*LookupError: '/redesSociales' does not exist*
*at app/anvil_extras/routing/_router.py:302*
*called from app/anvil_extras/routing/_router.py:343*
*called from app/anvil_extras/routing/_router.py:267*
*called from app/anvil_extras/routing/_router.py:151*
*called from app/anvil_extras/routing/_router.py:151*
*called from app/anvil_extras/routing/_router.py:111*
*called from app/anvil_extras/routing/_decorators.py:36*
*called from app/anvil_extras/routing/_router.py:235*
*called from app/anvil_extras/routing/_router.py:140*
*called from app/anvil_extras/routing/_router.py:140*
*called from app/anvil_extras/routing/_navigation.py:54*
#routing: navigation triggered: url_hash='/acercaDe', url_pattern='/acercaDe', url_dict={}
#routing: entering navigation level: 0
  #routing: checking templates and redirects
  #routing: unchanged template: 'RouterAreaPublica'
  #routing: RouterAreaPublica.on_navigation() called
  #routing: loading route: 'AcercaDe' from cache
#routing: exiting navigation level: 0
#routing: navigation complete
```

**Comment by s-cork** [comment, marked as answer]
From what I can see you've misunderstood the path part of the Template argument

```python

@routing.template(path='Marketing')
class RouterMarketing(RouterMarketingTemplate):
```
The above code says - my `RouterMarketing` template has all paths starting with `Marketing`

But the menu you build for the `RouterMarketing` form sets the hashes to (e.g.) `"/mercados"`, which should be `"Marketing/mercados"`. 

assuming that's what you want the URLs to look like then fixing your built menu should fix the navigation

If you want different URLs you'll need to think about things a little differently.
If that's the case let me know and we can look at this.

**Reply by RicardoFernandez-UY** [reply]
Yes!!!! 
You are 100% right. I completely misunderstood the use of the path. 
Now that you told me, I reviewed the docs, and I don't understand how I missed it. 
I corrected the menu items and other minor changes to the Form hashes, and it is working as expected. 
Thank you!!!
